### PR TITLE
osd: extend OMAP_GETKEYS and GETVALS to include a 'more' output field

### PR DESCRIPTION
**Pull request by liewegas**
The current omap get methods cannot communicate to the client that their
results were truncated due to hitting a limit on the number of keys or
bytes, making those server-side limits extremely dangerous to
enforce.

Extend the API to implement a new set of methods that explicitly expose a
"more" flag.

The old interfaces are marked deprecated.

Open question: should we change the old implementations to loop on the client
side?

### CODE REVIEW

**Comment by liewegas** [comment]
@jdurgin Should we make the old omap_get_{keys,vals} loop on the client?  I'm thinking yes...

**Comment by liewegas** [comment]
Okay, updated to do the looping on the client in the cases where we can.

This introduces a bunch of new warnings across the code base for various users of the deprecated API.  We can fix those before or after merge (I don't feel strongly either way).  I could also separate out the bit that marks them deprecated so that we don't "break" master with those warnings until they are fixed.  But I think having them there will be the right level of motivation to get them fixed...

The two critical users are rgw and the mds, @mattbenjamin @yehudasa @jcsp

File: src/librados/librados.cc [hunk 1]
```diff
@@ -266,7 +266,8 @@ void librados::ObjectReadOperation::omap_get_vals(
   int *prval)
 {
   ::ObjectOperation *o = &impl->o;
-  o->omap_get_vals(start_after, filter_prefix, max_return, out_vals, prval);
```
Comment (jdurgin) [inline src/librados/librados.cc]: looks like these librados test changes slipped into the wrong commit - they wouldn't build yet here

File: src/include/rados/librados.h [hunk 2]
```diff
@@ -3124,7 +3124,28 @@ CEPH_RADOS_API void rados_read_op_omap_get_vals(rados_read_op_t read_op,
...
+ * Start iterating over key/value pairs on an object.
+ *
+ * They will be returned sorted by key.
+ *
+ * @param read_op operation to add this action to
+ * @param start_after list keys starting after start_after
+ * @param filter_prefix list only keys beginning with filter_prefix
+ * @param max_return list no more than max_return key/value pairs
+ * @param iter where to store the iterator
+ * @param prval where to store the return value from this action
```
Comment (jdurgin) [inline src/include/rados/librados.h]: should describe pmore for each of these

File: src/include/rados/librados.h [hunk 3]
```diff
@@ -3124,7 +3124,28 @@ CEPH_RADOS_API void rados_read_op_omap_get_vals(rados_read_op_t read_op,
 				                const char *filter_prefix,
 				                uint64_t max_return,
 				                rados_omap_iter_t *iter,
-				                int *prval);
+				                int *prval)
+  __attribute__((deprecated)); /* use v2 below */
```
Comment (jdurgin) [inline src/include/rados/librados.h]: seems like the originals are useful convenience functions that shouldn't be deprecated now that they're looping on the client side

File: PendingReleaseNotes [hunk 4]
```diff
@@ -7,3 +7,13 @@
...
+
+ * The omap_get_keys and omap_get_vals librados functions have been deprecated
+   in favor of omap_get_vals2 and omap_get_keys2.  The new methods include an
+   output argument indicating whether there are additional keys left to fetch.
+   Previously this had to be inferred from the requested key count vs the
+   number of keys returned, but this breaks with new OSD-side limits on the
+   number of keys or bytes that can be returned by a single omap request.  These
+   limits were introduced by kraken but were effectively disabled (by setting a
+   very large limit of 1 GB) because users of the newly deprecated interface
+   cannot tell whether they should fetch more keys or not.
```
Comment (jdurgin) [inline PendingReleaseNotes]: mention looping for the originals, and imo no need to deprecate them, but warn about old clients + newer osd behavior here

File: src/librados/librados.cc [hunk 5]
```diff
@@ -336,19 +371,61 @@ void librados::ObjectReadOperation::is_dirty(bool *is_dirty, int *prval)
...
+    op.omap_get_vals2(start_after, filter_prefix, max_return, &out, &more,
+		      nullptr);
+    bufferlist bl;
+    int ret = operate(oid, &op, &bl);
+    if (ret < 0) {
+      return ret;
+    }
+    if (more) {
+      if (!out.empty()) {
+	return -EINVAL;  // wth
```
Comment (jdurgin) [inline src/librados/librados.cc]: what's  this case for?

File: src/librados/librados.cc (hunk 1 above)
Comment (liewegas) [inline src/librados/librados.cc]: these are calling the Objecter::ObjectOperation, not the librados structure

File: PendingReleaseNotes (hunk 4 above)
Comment (liewegas) [inline PendingReleaseNotes]: Adding warning.

The reason to deprecated them is to make sure the developer is aware they're using a fundamentally broken/unsafe interface...

File: src/librados/librados.cc (hunk 5 above)
Comment (liewegas) [inline src/librados/librados.cc]: it's just to guard the out.rbegin() dereference on the next line. it's not supposed to happen, but an assert would be rude.

File: src/include/rados/librados.h (hunk 2 above)
Comment (liewegas) [inline src/include/rados/librados.h]: fixed

File: src/include/rados/librados.h (hunk 3 above)
Comment (liewegas) [inline src/include/rados/librados.h]: These _read_op_* methods are used to build up a multi-op transaction.. we can't do magic client-side looping for them.

**Pull request merged**